### PR TITLE
Download geckodriver automatically

### LINKED ISSUE #14: When downloading latest coverage build, also download latest geckodriver version in the tools/ directory

### PR DESCRIPTION
**Pull request by rhcu**
Fixes #14.

Added functionality to `latest_cov_build.py` to download the latest version of Geckodriver to `tools/` folder, which should be created if doesn't exist. 
Differentiated between `.zip` Windows archives and `tar.gz` Unix archives. 

Tested on Linux Ubuntu 16.04.

### CODE REVIEW

File: latest_cov_build.py [hunk 1]
```diff
@@ -1,10 +1,25 @@
...
-    from urllib.request import urlretrieve
+    from urllib.request import urlopen, urlretrieve
+    from urllib.parse import urlencode
 except ImportError:
-    from urllib import urlretrieve
+    from urllib import urlencode, urlretrieve
+    from urllib2 import urlopen
+
+
+def get_json(url, params=None):
```
Comment (marco-c) [inline latest_cov_build.py]: Let's use the `requests` library, so we don't have to implement this ourselves.

File: latest_cov_build.py [hunk 2]
```diff
@@ -18,3 +33,51 @@
 with tarfile.open(name, 'r:bz2') as tar:
     tar.extractall()
 os.remove(name)
+
+# Geckodriver download
+# OS information for correct geckodriver version
+name = sys.platform
```
Comment (marco-c) [inline latest_cov_build.py]: No need to define a `name` variable, let's use `sys.platform` directly.

File: latest_cov_build.py [hunk 3]
```diff
@@ -18,3 +33,51 @@
...
+bitness = platform.architecture()[0]
+latest_version = get_json('https://api.github.com/repos/mozilla/geckodriver/releases/latest')
+tag_name = latest_version['tag_name']
+base_url = 'https://github.com/mozilla/geckodriver/releases/download/' + tag_name
+
+# Linux OS
+if name == 'linux':
```
Comment (marco-c) [inline latest_cov_build.py]: Instead of equality, this should check that the string starts with `linux` (e.g. on my machine sys.platform is `linux2`).

File: latest_cov_build.py [hunk 4]
```diff
@@ -18,3 +33,51 @@
...
+    if bitness == '64bit':
+        base_url += '/geckodriver-' + tag_name + '-linux64.tar.gz'
+        print(base_url)
```
Comment (marco-c) [inline latest_cov_build.py]: Nit: remove logging statements

File: latest_cov_build.py [hunk 5]
```diff
@@ -18,3 +33,51 @@
...
+    if bitness == '64bit':
+        base_url += '/geckodriver-' + tag_name + '-win64.zip'
+        print(base_url)
+    else:
+        base_url += '/geckodriver-' + tag_name + '-win32.zip'
+
+# Create 'tools/' directory if doesn't exist
+if not os.path.exists('tools'):
+    os.makedirs('tools')
+
```
Comment (marco-c) [inline latest_cov_build.py]: In the previous branches, you could store the file name instead of appending to base_url, then here you could do:
```Python
urlretrieve('{}/{}'.format(base_url, file_name), file_name)

if file_name.endswith('zip'):
    with zipfile...
elif file_name.endswith('tar.gz'):
    with tarfile...
```

This way we can avoid most of the differences between the .zip and the .tar.gz branches.

**Comment by marco-c** [comment]
Thanks! Looks good overall, just a few minor comments.

**Comment by rhcu** [comment]
Thanks for comments. I'll work on them. :)

**Comment by rhcu** [comment]
I've occasionally included coverage build file to commit and deleted it in the second.

**Comment by marco-c** [comment]
> I've occasionally included coverage build file to commit and deleted it in the second.

OK, no problem.
If you want, you can rewrite your Git history using `git rebase`.
If you use `git rebase -i HEAD~3`, it will allow you to change the history (using some operation, or changing the order of the commits) related to the latest three commits and you can squash (or fixup, to avoid merging commit messages) them into a single commit.

You don't always have to do it as I can do it myself from the GitHub interface, but I think it can be useful for you to try.

**Comment by rhcu** [comment]
Thanks. :) I have just added archives to .gitignore, and can try `git rebase` now.

File: .gitignore [hunk 6]
```diff
@@ -5,7 +5,8 @@ __pycache__/
 
 # C extensions
 *.so
-
+# tar.gz
+*.tar.bz2
```
Comment (marco-c) [inline .gitignore]: Let's do this in a separate PR, we should make each PR self-contained (also, when you do, please put it at the bottom of the file as that's where ignored files specific to this projects are).

Comment (rhcu) [reply .gitignore]: Done. :)

**Comment by marco-c** [comment]
Thanks ;)

Do you want to try the rebasing with `git rebase`? Now you have 6 commits, so you should do `git rebase -i HEAD~6`.

**Comment by marco-c** [comment]
BTW, for the following PRs, I suggest you create a different branch for each feature/bugfix you work on, as it will be more flexible (this way, you can work on different things in parallel on diffferent branches).
Short explanation here: https://guides.github.com/introduction/flow/.

**Comment by rhcu** [comment]
Thanks. Yup, I'll try to squash them into one. :)

**Comment by rhcu** [comment]
Seems, that they are all squashed into one now. I'll try to create new branches in the next PRs.

**Comment by marco-c** [comment]
There are still 8 in this PR, have you run `git push`? (in this case probably you will need `git push --force` as you have rewritten the history)

**Comment by marco-c** [comment]
After you rebase, you shouldn't run `git pull`, otherwise it will re-apply the old commits that you already had on master.

Basically, you have this history now both on your local copy and on your github repo:
COMMIT1
COMMIT2
COMMIT3
COMMIT4
COMMIT5
COMMIT6
COMMIT7
COMMIT8

If you do `git rebase -i HEAD~8` and squash them into one, you will end up with this history on your local copy:
NEW_COMMIT1

And this history on your github repo:
COMMIT1
COMMIT2
COMMIT3
COMMIT4
COMMIT5
COMMIT6
COMMIT7
COMMIT8

Now, if you do `git pull`, the commits from your github repo will be applied in your local copy, so you will end up with 10 commits (the new squashed one, the old 8 ones, plus a merge commit). Instead, you want your local copy to overwrite the github repo, so you need to run `git push --force`.
This way, both your local copy and your github repo will have the history:
NEW_COMMIT1

**Comment by rhcu** [comment]
Thanks for explanation! Seems that it worked. :)
